Attempt at refactoring & decoupling the OnChainHeadState logic

KtorZ opened the pull request:
## History

- :round_pushpin: **Cleanup Direct.Tx module from previous approach.**
    We discussed that, rather to be in that direction, the dependency should be the other way around. That is, we want the 'State' module to sit in between the direct-chain component and the low-level transaction construction. Hence, it is the 'State' module which will use the 'Direct.Tx' module down the line to construct and observe the transitions.

- :round_pushpin: **Type-drive refactoring of the Hydra.Chain.Direct module.**
     We start from the type, which outlines the interface that we want. And from there, write implementation. It's a very top-down approach where we goes more and more into details with the goal of making the module hierarchy a clearer: that is, the Chain.Direct is fairly ignorant to (a) what's the internal state and (b) how it is represented. This allows to write functions in a generic way, to decouple from implementation details and therefore, make testing easier (since bottom parts can be tested independently and in depth, while upper layers can focus on integration only).

- :round_pushpin: **re-implement 'fromPostChainTx' using the new 'State' interface.**
    Much cleaner :D. And now, we can test each transition independently, which is great (although still yet to do :grimacing:).

- :round_pushpin: **re-implement 'runOnChainTx' in the Chain.Direct module.**
    This is now in principle, all that is needed in the Chain.Direct module which is now very generic about the state and the transitions. It can observe the chain and keep track of /some state/ but the representation of that state is unknown to it (implementation detail of the lower layers).

- :round_pushpin: **Cleanup Hydra.Chain.Direct.**
  
- :round_pushpin: **re-implement initialize, abort, commit, collect, close and fanout as independently testable functions.**
    Note that they should not be in STM, as per note.

- :round_pushpin: **re-implement 'observeTx' for all transitions.**
  
- :round_pushpin: **Revert putting TinyWallet in Direct.State.**

## Remarks

- Tests don't compile yet since I _slightly_ changed the interface of the low-level tx construction functions (which no longer return an `OnChainHeadState` but instead, returns _observations_. Need to be done, but not if we end up dropping this whole approach. 

- This refactoring does not simplify anything about the low-level tx construction or observation. Rather, it introduces an indirection between those low-level details and the higher direct-chain component which is now manipulating _some_ opaque state. This buys us mainly two things:
   -  It gives us more freedom in changing the internal representation of the on-chain head state, without impacting the rest of the code; 
   - It makes the state transitions function testable independently!  

- From there, a next logical step is to get rid of the `TVar` storing the on-chain state, and move it inside the head logic. We can achieve this by yielding new states as part of `OnChainTx` events and, providing it back as part of `PostChainTx` commands, which effectively make the chain-component stateless (the state is within the requests/responses!). 

- In parallel, we can also work on simplifying the low-level building and observing of transactions by introducing an embedded DSL similar to the Plutus constraints. the main guiding principle for the DSL should be that we ought to be able to write one description of each transaction and either interpret it to build a transaction from a known context or, to observe it from a given tx. This would reduce a lot of the duplications between construction/observation and gives us one declarative and consistent approach for both constructing and observing.

- Finally, once the on-chain state is part of the head logic's state, we can consider representing the head logic not as an aggregate state, but as sequence of states instead (folding only on-demand) to cope with rollbacks in an easy way!

KtorZ committed:
Attempt at refactoring & decoupling the OnChainHeadState logic

  This commit / PR is mostly here for fueling the discussion and
  illustrating what we've been discussing recently; namely: being able
  to decouple the construction and observation of Hydra transactions
  from low-level details and state management.

  Here is defined a new module 'Hydra.Chain.Direct.State' which exposes
  an opaque 'OnChainHeadState' with state-machine's states being tracked
  with a kind witness. The module then defines an API which takes care
  of constructing and extracting details that depend on previously
  observed events in the Hydra state-machine lifecycle. This simplifies
  greatly the construction and observation of on-chain transactions,
  and, offers a more granular API which can be tested independently
  (given some state, we should able to assert specific properties w.r.t
  to what's observable and what isn't).

  So far, only the commitTx and observeCommitTx are covered to give a
  sense of what this could look like. Note also that the current
  internal representation of a state may not be optimal, but it is meant
  as being a drop-as-much-as-possible-in replacement for our current
  implementation. Then, because it is actually opaque to other modules,
  we have rooms for refactoring and changing how we internally keep
  track of the on-chain state (e.g. by storing a sequence of
  transactions rather than sum of products).

ch1bo commented:
We discussed the approach and how it could fit a bit better "in-between" `Hydra.Chain.Direct` and `Hydra.Chain.Direct.Tx` modules. Likely we want to start with step 1 on this doodle:

![image](https://user-images.githubusercontent.com/2621189/154663938-a4eb6895-aeb1-4d3e-ae13-3f746c99753f.png)

KtorZ committed:
Cleanup Direct.Tx module from previous approach.

  We discussed that, rather to be in that direction, the dependency should be the other way around. That is, we want the 'State' module to sit in between the direct-chain component and the low-level transaction construction. Hence, it is the 'State' module which will use the 'Direct.Tx' module down the line to construct and observe the transitions.

KtorZ committed:
Type-drive refactoring of the Hydra.Chain.Direct module.

   We start from the type, which outlines the interface that we want. And from there, write implementation. It's a very top-down approach where we goes more and more into details with the goal of making the module hierarchy a clearer: that is, the Chain.Direct is fairly ignorant to (a) what's the internal state and (b) how it is represented. This allows to write functions in a generic way, to decouple from implementation details and therefore, make testing easier (since bottom parts can be tested independently and in depth, while upper layers can focus on integration only).

KtorZ committed:
re-implement 'fromPostChainTx' using the new 'State' interface.

  Much cleaner :D. And now, we can test each transition independently, which is great (although still yet to do :grimacing:).

KtorZ committed:
re-implement 'runOnChainTx' in the Chain.Direct module.

  This is now in principle, all that is needed in the Chain.Direct module which is now very generic about the state and the transitions. It can observe the chain and keep track of /some state/ but the representation of that state is unknown to it (implementation detail of the lower layers).

KtorZ committed:
Cleanup Hydra.Chain.Direct.

KtorZ committed:
re-implement initialize, abort, commit, collect, close and fanout as independently testable functions.

  Note that they should not be in STM, as per note.

KtorZ committed:
re-implement 'observeTx' for all transitions.

KtorZ committed:
Revert putting TinyWallet in Direct.State.

  This was a leaky abstraction anyway since the wallet was still needed outside of the state management anyway to finalize transactions. Plus, the wallet was only truly used to get a seed input from the UTXO for the initializeTx case, pervasively requiring STM everywhere as a consequence. The seed input selection is now done outside of the State module, which is kept focused on being a high-level bridge to the lower Direct.Tx one.

KtorZ committed:
Remove now-dead code.

KtorZ commented on hydra-node/src/Hydra/Chain/Direct/State.hs:
```diff
@@ -0,0 +1,396 @@
+take2Of4 (a, b, _c, _d) = (a, b)
+
+take2Of3 :: (a, b, c) -> (a, b)
+take2Of3 (a, b, _c) = (a, b)
```
I love Haskell.

KtorZ committed:
Fix TxSpec and the like to accomodate the new observeXXXTx interface.

  Note that I had to remove one test, which was testing the now-removed 'observeCommit' function. However, I feel a bit guilty now because that piece of logic has moved to the 'Hydra.Chain.Direct.State' module, and it therefore now untested. Thus, I feel obliged to write tests for that module now. Or at least, one test to cover back this bit of logic.

KtorZ committed:
Add some helpers to hydra-cardano-api to access input sets from Tx and UTxO.

KtorZ committed:
Move 'genTxIn' to Hydra.Ledger.Cardano (from Wallet's test code).

  This generator is actually useful in other contexts.

KtorZ committed:
Re-instate removed test on obsolete observeCommit

  Now tested as part of the 'observeTx' transition from a StInitialized.

ch1bo commented on hydra-node/src/Hydra/Chain/Direct.hs:
```diff
@@ -207,21 +218,19 @@ instance Exception ConnectException
             { localChainSyncProtocol =
                 InitiatorProtocolOnly $
-                  let peer = chainSyncClientPeer $ chainSyncClient tracer networkMagic callback party headState
+                  let peer = chainSyncClientPeer $ chainSyncClient tracer callback headState
```
Where did the `networkMagic` and `party` go? The latter was required to `observeInitTx` iirc

KtorZ replied:
They're now in the state itself.

ch1bo commented on hydra-node/src/Hydra/Chain/Direct/State.hs:
```diff
@@ -0,0 +1,400 @@
+-- I'd suggest to remove the `MonadThrow` instance and instead have the
+-- potentially failing functions return an 'Either' and, do the exception
+-- throwing in the upper-layer, wrapping the error with details such as the
+-- wallet UTxO.
```
This NOTE/TODO seems out of date?

KtorZ replied:
Ah yes. I did that already :sweat_smile: !

ch1bo commented on hydra-node/src/Hydra/Chain/Direct/State.hs:
```diff
@@ -0,0 +1,400 @@
+  observeTx ::
+    Tx ->
+    OnChainHeadState st ->
+    Maybe (OnChainTx Tx, SomeOnChainHeadState)
```
Do we even require a type class after all? `OnChainHeadState st` is still a single type and wouldn't we be able to implement `observeTx :: Tx -> OnChainHeadState st -> Maybe (OnChainTx Tx, SomeOnChainHeadState)` with something like the following?

```hs
observeTx tx OnChainHeadState{networkId, ownParty, stateMachine} =
  case stateMachine of
    Idle{} ->
      (event, observation) <- observeInitTx networkId ownParty tx
      -- etc
    Initialized{initialInitials, initialCommits} ->
      observeCommit <|> observeCollectCom <|> observeAbort
      -- etc
-- ... etc using all the same code in the individual instances
```

KtorZ replied:
Yes and no, I had the type-class originally because the instances were fully capturing the transitions (see #228). But, this may be _too much dependent type_. 

Still, having a type-class here allows for overloading `observeTx` in its arguments. Without that, we would need to pattern-match on the call-site to figure out in which state we are, and call the right transition function, and put the state back in an existential. That's perhaps doable (and actually, a good in-between proposition between the current state of affairs and #228 :sweat_smile: ...)
